src/mysql-util/src/tunnel.rs

```diff
@@ -21,8 +25,45 @@ use crate::MySqlError;
 pub enum TunnelConfig {
     /// Establish a direct TCP connection to the database host.
     Direct,
-    // TODO: Implement SSH tunneling for MySQL connections
-    // TODO: Implement AWS PrivateLink tunneling for MySQL connections
+    /// Establish a TCP connection to the database via an SSH tunnel.
+    /// This means first establishing an SSH connection to a bastion host,
+    /// and then opening a separate connection from that host to the database.
+    /// This is commonly referred by vendors as a "direct SSH tunnel", in
+    /// opposition to "reverse SSH tunnel", which is currently unsupported.
+    Ssh { config: SshTunnelConfig },
+    /// Establish a TCP connection to the database via an AWS PrivateLink
+    /// service.
+    AwsPrivatelink {
+        /// The ID of the AWS PrivateLink service.
+        connection_id: GlobalId,
+    },
+}
+
+/// A MySQL connection with an optional SSH tunnel handle.
+///
+/// This wrapper is intended to be used in place of `mysql_async::Conn` to
+/// keep the SSH tunnel alive for the lifecycle of the connection by holding
+/// a reference to the tunnel handle.
+#[derive(Debug)]
+pub struct MySqlConn {
+    pub conn: Conn,
+    _ssh_tunnel_handle: Option<ManagedSshTunnelHandle>,
+}
+
+impl MySqlConn {
+    pub fn inner_mut(&mut self) -> &mut Conn {
```

Comment:
I'd implement `Deref` and `DerefMut`

Reply:
woah that completely cleaned things up! Super cool - I didn't know that this makes it implicitly inherit the mutable methods on the target type!

Reply:
its great!